### ISSUE TITLE
Implement lightweight networking failure injection

### ISSUE THREAD

**Issue by seeker89** [powerfulseal/powerfulseal#285]
# TL;DR

> Add a simple way to inject lightweight network latency with a small, controlled blast-radius, as well as no SSH access.

We've [been talking about this one for a while](https://github.com/bloomberg/powerfulseal/issues/136), so I think it's time for this to materialize.

## Use case

With the benefit of hindsight, I think that the main use case that we want to cover (and cover it well) is the following:

> I have a deployment running my app, and a service that routes to the pods of that deployment. Optionally, I might be using a service mesh to have more flexible routing solution.

> I want to leave my existing pods running, but introduce an extra few copies that will have networking problems injected into them. 

> I then want to get some part of my traffic routed to that deployment (for example for it to get picked up by the service, or by the service mesh).

Right now, this can already be done, but it's a manual process. It looks like this:

- create a new deployment that mirrors some existing deployment
  - compose the necessary yaml
  - set the labels on the new deployment in a way that it will be routed to with whatever mechanism you're using
  - use the `kubectl` functionality to deploy that
- inject the failure
  - use something like https://github.com/shopify/toxiproxy
  - configure it to do the kind of failure that you're after
  - configure the app so that it uses the proxy
- do the validation
  - this is typically through an `HTTPProbe`

So it's possible, but pretty labour-intense and pretty static.

In this issue, I'd like to suggest we cover this popular use case, so that it becomes easy to use. So there are three separate concerns:

1. cloning (& tweaking)
2. fault injection
3. routing

### 1. Cloning

Let's assume that you are either using a kubernetes service for routing your stuff, or that whatever else you're using is relying on the services. You might have a setup like this. A deployment:

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: powerfulseal
  namespace: powerfulseal
spec:
  replicas: 1
  selector:
    matchLabels:
      name: powerfulseal
  template:
    metadata:
      labels:
        name: powerfulseal
    spec:
      containers:
      - ports:
        - containerPort: 8080
          name: powerfulseal
          protocol: TCP
        ...
```

And then a matching service, that uses the `name=powerfulseal` label:

```yaml
apiVersion: v1
kind: Service
metadata:
  name: powerfulseal
  namespace: powerfulseal
spec:
  ports:
  - name: powerfulseal
    port: 8080
    protocol: TCP
    targetPort: powerfulseal
  selector:
    name: powerfulseal
  type: ClusterIP
...
```

[A](https://github.com/kubernetes/kubernetes/issues/24946) [while](https://github.com/kubernetes/kubernetes/issues/24152) [back](https://github.com/kubernetes/kubernetes/issues/14961#issuecomment-183431648) having deployments with labels that are a superset of one another was a problem. Technically we could get away with having the same set, but it would be pretty confusing.

Instead, I suggest we let the user pick the service in question, let the seal copy the labels, and add an extra one, for increased visibility.

So in the examples above, it would calculate the label set to be, for example:

```yaml
    metadata:
      labels:
        name: powerfulseal
        chaos: true
```

### Fault injection

I would like to provide a system where it's easy to add custom modifications. There are existing proxy libraries, like [toxiproxy](https://github.com/shopify/toxiproxy) that we can leverage using a sidecar. But I'd like to keep it pretty open, so that we can add more mutation options later on.

### Routing

For now, I think that we can just rely on the services. Ingresses, as well as many discovery solutions will use that, so it should provide most of the value.

## Feature

So here's a high-level idea of what I'm picturing:

```yaml
clone:

  # find the source to copy from
  source:

    # most of the time it'll be a deployment, so let's target this as the first one
    deployment:
      name: some-service
      namespace: some-namespace

  # what labels to use in the clone (so that your service can pick them up)
  # additional to all of these, we'll add an extra set: `chaos=true`
  # if this is not specified, it will just default to the ones from the source + the extra one
  labels:

    # possibly copy from a service
    - service:
        name: some-service
        namespace: some-namespace

    # or add static ones, to match what you need
    - static:
        key1: value1
        key2: value2


  # a list of mutations to apply, like injecting latency through a sidecar with toxiproxy, and later others
  mutate:

    # package pre-configured toxiproxy in a side car
    - toxiproxy:

      # whether the other containers should get their `http{,s}_proxy` variables overwritten to write here
      override_clone_proxy_env: true

      # whether the proxy container should copy proxy env from the source container, if any
      copy_proxy_env_from_source: true

      # list of toxiproxy toxics to implement
      toxics:

        - latency: {}
        - down: {}
        - bandwidth: {}

    # or add a sidecar container with tc
    - tc:

      # specify arguments for tc to execute
      args: []

      # or use one of the built-in presets
      preset:
        latency: {}

      # in case you want to use a custom container with tc built-in
      image: docker.io/mytccontainer:v1.0.0


    # this modifies the env by overriding variables with specific values
    - environment:

      # for example, if your application has some chaos engineering built-in, you can activate it
      - name: CHAOS_ACTIVE
        value: true

      # otherwise, if you're running some external failure proxy, you could do something like 
      - name: CHAOS_ACTIVE
        value: true

      # if you're on the TCP level, you might need to modify some variables to point to the proxy on a different port
      - name: REDIS_PORT
        value: 6378
      - name: REDIS_HOST
        value: localhost
```

Anyone, that's my current braindump. Keen for comments!

**Comment by dgzlopes** [comment]
I think this is the right path :)

Some ideas, questions, proposals, nits, things:

- I think too that targeting a deployment is enough for now.
- On the labels section, the `service` part means that we want to copy the labels from another service?
- I'm not sure how I feel about the `mutate` section:
  + Maybe we can keep this tool-agnostic? Or add some pre-defined templates/abstraction around it? I'm thinking about defining a finite set of network attacks/mutations e.g packet loss, latency... It's more than enough for most use cases and you only have to learn one tool (Powerfulseal). But I understand that for some use-cases it's better to tweak the underlying proxy/tool... Mixed thoughts.
- What happens when we finish the injection? Do we delete the clones? What if I want to run an injection for some minutes - wait - and do another injection? Maybe we can add one flag for `persistence` or something like that.

**Comment by seeker89** [comment]
> I think this is the right path :)
> 
> Some ideas, questions, proposals, nits, things:
> 
> * I think too that targeting a deployment is enough for now.
> * On the labels section, the `service` part means that we want to copy the labels from another service?

That's what I was thinking - copy the labels that the actual service item uses to match pods.

> * I'm not sure how I feel about the `mutate` section:
>   
>   * Maybe we can keep this tool-agnostic? Or add some pre-defined templates/abstraction around it? I'm thinking about defining a finite set of network attacks/mutations e.g packet loss, latency... It's more than enough for most use cases and you only have to learn one tool (Powerfulseal). But I understand that for some use-cases it's better to tweak the underlying proxy/tool... Mixed thoughts.

When I first wrote it up, it was a generic thing. But they I changed my mind, for these reasons:

- it's open source software, no need to hide what's running underneath
- to use this feature effectively, you're still probably going to need some kind of idea of what the thing underneath is doing
- there is existing documentation that can be leveraged
- give the credit where it's due
- there will probably be different tools that can provide the same functionality (we could implement this with a sidecar container that has 'tc add... && sleep X && tc delete .... && sleep forever` in it to add latency too). So it would like make sense to just call things by their names.

> * What happens when we finish the injection? Do we delete the clones? What if I want to run an injection for some minutes - wait - and do another injection? Maybe we can add one flag for `persistence` or something like that.

That's an interesting idea. We could definitely consider it, although it looks like a nice-to-have, rather than a must-have in the first version of this.

**Comment by seeker89** [comment]
I updated the initial comment with a little example of what I meant about different tools and tc.

**Comment by karthequian** [comment]
I like this- it makes sense to have seal copy the labels like you've mentioned in this section in the feature:

```
    # possibly copy from a service
    - service:
        name: some-service
        namespace: some-namespace
```

I did not understand what the alternate proposal meant where you said this. Does that mean add additional k/v pairs to an existing service and then find these in the clone?
```
    # or add static ones, to match what you need
    - static:
        key1: value1
        key2: value2
```

I like that the mutate section has different stanzas for implementations- it makes sense to me, and maybe pick tc or toxiproxy to inject issues. It could be generalized, by having an interface for mutations maybe, and then different implementations for toxiproxy/tc/whatever, but agree that maybe it's overkill. I think as long as it's okay to extend powerfulseal with more kinds of mutates from others, the format should be fine, if not consider generalizing it.

**Comment by seeker89** [comment]
> I did not understand what the alternate proposal meant where you said this. Does that mean add additional k/v pairs to an existing service and then find these in the clone?

I meant that in order for the new deployment to be picked up by the service, we can copy the labels that the service matches on, and add an extra one just for differentiation, like `chaos: true` or something like that.

**Comment by seeker89** [comment]
Hey all, I've put together a first draft of what I meant with this issue. Have a read and tell what you think https://github.com/powerfulseal/powerfulseal/pull/302
